A11y service impl

matthew-hagemann opened the pull request:
closes #311

d-loose commented on provd/protos/accessibility.proto:
```diff
@@ -4,205 +4,53 @@ package accessibility;
+    // Zoom
+    rpc GetDesktopZoon(google.protobuf.Empty) returns (google.protobuf.BoolValue) {}
+    rpc EnableDesktopZoon(google.protobuf.Empty) returns (google.protobuf.Empty) {}
+    rpc DisableDesktopZoon(google.protobuf.Empty) returns (google.protobuf.Empty) {}
```
```suggestion
    rpc GetDesktopZoom(google.protobuf.Empty) returns (google.protobuf.BoolValue) {}
    rpc EnableDesktopZoom(google.protobuf.Empty) returns (google.protobuf.Empty) {}
    rpc DisableDesktopZoom(google.protobuf.Empty) returns (google.protobuf.Empty) {}
```

matthew-hagemann replied:
hehe, okay, i've installed a spellchecked :sweat_smile:

matthew-hagemann commented on provd/internal/services/accessibility/accessibility.go:
```diff
@@ -0,0 +1,121 @@
+
+func getBooleanSettings(boolSettings booleanGetter, key string) bool {
+	return boolSettings.GetBoolean(key)
+}
```
There is no lines of code gain using a getter, but I thought it looked more consistent along side the setter where we do make those gains.

didrocks commented on provd/internal/services/user/actions:
```diff
@@ -0,0 +1,34 @@
+Accounts.FindUserByName(name: mock-user)
```
hum, I think that was committed by mistake?

didrocks commented on provd/internal/services/timezone/testdata/TestGetTimezone/golden/error_on_empty_request:
```diff
@@ -0,0 +1 @@
+<nil>
```
See, we shouldn’t have this golden file.

didrocks commented:
@matthew-hagemann are you waiting for me on that one? It doesn’t seem my comments are addressed or acked (and you didn’t request for review).

Just asking, fine if you didn’t have time, just to ensure that you didn’t miss them :)

matthew-hagemann commented:
> @matthew-hagemann are you waiting for me on that one? It doesn’t seem my comments are addressed or acked (and you didn’t request for review).
> 
> Just asking, fine if you didn’t have time, just to ensure that you didn’t miss them :)

Hi @didrocks! Sorry I didn't ack, and ty for the review! We shifted focus last week to packaging and prepping for demos, so I haven't had a chance to circle back to features yet, but will address these first as soon as I do :slightly_smiling_face:

didrocks commented:
> Hi @didrocks! Sorry I didn't ack, and ty for the review! We shifted focus last week to packaging and prepping for demos, so I haven't had a chance to circle back to features yet, but will address these first as soon as I do 🙂

No worry, I was just wondering if that was an oversight :) Good luck with the demo!

matthew-hagemann commented on provd/internal/services/accessibility/accessibility.go:
```diff
@@ -0,0 +1,121 @@
+		return nil, status.Errorf(codes.Internal, "failed to connect to location settings")
+	}
+
+	isWritable = s.keyboardSettings.IsWritable("sticky-keys")
```
Thanks! At the moment I don't think I'll add anymore, I just picked one per `GSettings` object as a "ping" test, but will keep this in mind :slightly_smiling_face:

matthew-hagemann commented on provd/internal/services/accessibility/accessibility_test.go:
```diff
@@ -0,0 +1,169 @@
+
+func (g gSettingsSubsetMock) GetDouble(key string) float64 {
+	if g.getDoubleError {
+		return 0
```
My understanding of the underlying func this calls is that is will just return the default value if somthing goes wrong:
```
func (this0 *Settings) GetDouble(key0 string) float64 {
	var this1 *C.GSettings
	var key1 *C.char
	if this0 != nil {
		this1 = (*C.GSettings)(this0.InheritedFromGSettings())
	}
	key1 = _GoStringToGString(key0)
	defer C.free(unsafe.Pointer(key1))
	ret1 := C.g_settings_get_double(this1, key1)
	var ret2 float64
```
It doesn't return a success bool so we can't tell if there was a failure?

didrocks replied:
Exactly! That was my point: the test is lying telling that we error out :) So the idea is really to set to some value and check the return value. If there are 2 values, try both :) but don’t tell there is an error we can trigger (because we can from the mock, but not in reality).

Consequently, the mock should not allow to fail either to reflect the real behaviour.

matthew-hagemann commented on provd/internal/services/accessibility/hearing_test.go:
```diff
@@ -0,0 +1,157 @@
+		"Return visual alerts are enabled":  {wantTrue: true},
+
+		// Error cases
+		"Error when can't get state of visual alerts": {getBooleanError: true},
```
I have a similar issue to the one I have with `GetBoolean`, there is no success bool returned, only the default value, on error:

```
func (this0 *Settings) GetBoolean(key0 string) bool {
	var this1 *C.GSettings
	var key1 *C.char
	if this0 != nil {
		this1 = (*C.GSettings)(this0.InheritedFromGSettings())
	}
	key1 = _GoStringToGString(key0)
	defer C.free(unsafe.Pointer(key1))
	ret1 := C.g_settings_get_boolean(this1, key1)
	var ret2 bool
```

didrocks replied:
see previous reply.

didrocks commented on provd/internal/services/testdata/TestRegisterGRPCServices/golden:
```diff
@@ -126,10 +126,10 @@ privacy.PrivacyService:
 timezone.TimezoneService:
     methods:
-        - name: SetTimezone
+        - name: GetTimezone
```
hum, I am puzzled about the order (and the reordering of all those calls), are the golden file reproductible, always printing the same state order everytime?

matthew-hagemann replied:
No, they actually update the order every time I see.

Here are two outputs of running with `--update`, one after the other

```
diff --git a/provd/internal/services/testdata/TestRegisterGRPCServices/golden b/provd/internal/services/testdata/TestRegisterGRPCServices/golden
index 80ed5885..cf3b1da0 100644
--- a/provd/internal/services/testdata/TestRegisterGRPCServices/golden
+++ b/provd/internal/services/testdata/TestRegisterGRPCServices/golden
@@ -1,107 +1,107 @@
 accessibility.AccessibilityService:
     methods:
-        - name: GetScreenKeyboard
+        - name: GetVisualAlerts
           isclientstream: false
           isserverstream: false
-        - name: GetStickyKeys
+        - name: EnableVisualAlerts
           isclientstream: false
           isserverstream: false 
...
```

```
diff --git a/provd/internal/services/testdata/TestRegisterGRPCServices/golden b/provd/internal/services/testdata/TestRegisterGRPCServices/golden
index 80ed5885..bb1c558c 100644
--- a/provd/internal/services/testdata/TestRegisterGRPCServices/golden
+++ b/provd/internal/services/testdata/TestRegisterGRPCServices/golden
@@ -1,107 +1,107 @@
 accessibility.AccessibilityService:
     methods:
-        - name: GetScreenKeyboard
+        - name: GetHighContrast
           isclientstream: false
           isserverstream: false
-        - name: GetStickyKeys
+        - name: GetSlowKeys
           isclientstream: false
           isserverstream: false
...
```

didrocks replied:
Yeah, so the order is not deterministic. It’s definitively something that has to be fixed/figured out before merging this, or you will go crazy on every update :)

As we calling them in a non deterministic order in those tests? Is it something expected and only what matters is what get called in terms of how many times they are called with each parameters?
If so, we can have some reorder function that we intercept before saving in the golden file.

didrocks reviewed (changes requested):
And here we go! Thanks for all the fixes.

The last 2 remaining comments are:
- check the golden file actions are in the correct order and reproducible (always return the same one so that we don’t start getting flaky tests)
- rework the tests + mock to not make people believe we can get an error on the `Get*` functions.

Good work! I think we are close to merging this one :)

matthew-hagemann commented:
@didrocks I've updated the `Get*` tests and removed false error cases as well as add a test for when a default double is returned :slightly_smiling_face:

didrocks commented:
Excellent! (unsure why there is one of the comment I can’t resolve). So the last thing to figure out is the golden file update to always save the call in a deterministic way. (see my comment on the opened conversation), correct?

didrocks reviewed (approved):
Once you removed the printf!

d-loose reviewed (approved):
LGTM :+1:
